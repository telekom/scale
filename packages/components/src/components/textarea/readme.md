# scale-textarea



<!-- Auto Generated Below -->


## Properties

| Property      | Attribute     | Description                             | Type                                              | Default     |
| ------------- | ------------- | --------------------------------------- | ------------------------------------------------- | ----------- |
| `cols`        | `cols`        | (optional) textarea column              | `number`                                          | `undefined` |
| `counter`     | `counter`     | (optional) Input counter                | `boolean`                                         | `undefined` |
| `disabled`    | `disabled`    | (optional) Input disabled               | `boolean`                                         | `undefined` |
| `helperText`  | `helper-text` | (optional) Input helper text            | `string`                                          | `''`        |
| `inputId`     | `input-id`    | (optional) Input checkbox id            | `string`                                          | `undefined` |
| `label`       | `label`       | (optional) Input label                  | `string`                                          | `''`        |
| `maxLength`   | `max-length`  | (optional) Input max length             | `number`                                          | `undefined` |
| `minLength`   | `min-length`  | (optional) Input min length             | `number`                                          | `undefined` |
| `name`        | `name`        | (optional) Input name                   | `string`                                          | `''`        |
| `placeholder` | `placeholder` | (optional) Input placeHolder            | `string`                                          | `''`        |
| `readonly`    | `readonly`    | (optional) Input readonly               | `boolean`                                         | `undefined` |
| `required`    | `required`    | (optional) Input required               | `boolean`                                         | `undefined` |
| `resize`      | `resize`      | (optional) textarea resize              | `"horizontal" \| "none" \| "unset" \| "vertical"` | `undefined` |
| `rows`        | `rows`        | (optional) textarea row                 | `number`                                          | `undefined` |
| `status`      | `status`      | (optional) Input status                 | `string`                                          | `''`        |
| `styles`      | `styles`      | (optional) Injected CSS styles          | `string`                                          | `undefined` |
| `transparent` | `transparent` | (optional) input background transparent | `boolean`                                         | `undefined` |
| `value`       | `value`       | (optional) Input value                  | `number \| string`                                | `''`        |


## Events

| Event          | Description                             | Type                                  |
| -------------- | --------------------------------------- | ------------------------------------- |
| `scaleBlur`    | Emitted when the input loses focus.     | `CustomEvent<void>`                   |
| `scaleChange`  | Emitted when the value has changed.     | `CustomEvent<InputChangeEventDetail>` |
| `scaleFocus`   | Emitted when the input has focus.       | `CustomEvent<void>`                   |
| `scaleInput`   | Emitted when a keyboard input occurred. | `CustomEvent<KeyboardEvent>`          |
| `scaleKeyDown` | Emitted on keydown.                     | `CustomEvent<KeyboardEvent>`          |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
