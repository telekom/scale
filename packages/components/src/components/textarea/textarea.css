/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

scale-textarea {
  --transition: all var(--telekom-motion-duration-transition)
    var(--telekom-motion-easing-standard);
  --radius: var(--telekom-radius-standard);
  --border: var(--telekom-spacing-unit-x025) solid
    var(--telekom-color-ui-extra-strong);
  --border-error: var(--telekom-spacing-unit-x05) solid
    var(--telekom-color-functional-danger-standard);
  /* FIXME:TOKEN replace with ui border hover */
  --border-color-hover: var(--telekom-color-primary-hovered, #f90984);
  --border-color-focus: var(--telekom-color-primary-pressed, #f90984);
  --border-color-disabled: var(--telekom-color-ui-disabled);
  --box-shadow-focus: 0 0 0 var(--telekom-spacing-unit-x05)
    var(--telekom-color-functional-focus);
  --spacing-x: var(--telekom-spacing-unit-x3);
  --color-disabled: var(--telekom-color-text-and-icon-disabled);
  --background-disabled: transparent;
  --border-color-readonly: var(--telekom-color-ui-subtle);
  --background-readonly: var(--telekom-color-ui-subtle);

  /*meta*/
  --font-weight-meta: var(--telekom-typography-font-weight-bold);
  --font-size-meta: var(--telekom-typography-font-size-small);
  --line-height-meta: var(--telekom-typography-line-spacing-standard);
  --spacing-y-meta: var(--telekom-spacing-unit-x1);
  --color-meta: var(--telekom-color-text-and-icon-standard);
  --color-meta-error: var(--telekom-color-text-and-icon-functional-danger);

  /*control*/
  --spacing-control: var(--telekom-spacing-unit-x6) var(--spacing-x)
    var(--telekom-spacing-unit-x3) calc(var(--spacing-x) - 1px);
  --transition-control: var(--transition);
  --font-size-control: var(--telekom-typography-font-size-body);
  --background-control: var(--telekom-color-ui-base);
  /*counter*/
  --transition-counter: var(--transition);
  --font-size-counter: var(--font-size-meta);
  --line-height-counter: var(--line-height-meta);
  --color-counter-error: var(--color-meta-error);

  /*helper-text*/
  --transition-helper-text: var(--transition);
  --font-size-helper-text: var(--font-size-meta);
  --line-height-helper-text: var(--line-height-meta);
  --color-helper-text: var(--telekom-color-functional-informational-standard);
  --color-helper-text-error: var(--color-meta-error);

  /*placeholder*/
  --transition-placeholder: var(--transition);
  --color-placeholder: var(--telekom-color-text-and-icon-additional);

  /*label*/
  --color-label: var(--telekom-color-text-and-icon-additional);
  --color-label-readonly: var(--telekom-color-text-and-icon-additional);
  /* FIXME:TOKEN z-index token is needed */
  --z-index-label: var(--scl-z-index-10);
  --transition-label: var(--transition);
  --font-size-label: var(--telekom-typography-font-size-body);
  --font-weight-label: var(--telekom-typography-font-weight-medium);
  --font-size-label-focus: var(--telekom-typography-font-size-footnote);
  --font-weight-label-focus: var(--telekom-typography-font-weight-bold);
}

.textarea {
  position: relative;
}

.textarea .textarea__control {
  width: 100%;
  margin: 0;
  resize: vertical;
  display: flex;
  outline: none;
  padding: var(--spacing-control);
  z-index: 1;
  box-sizing: border-box;
  transition: var(--transition-control);
  font-family: inherit;
  font-size: var(--font-size-control);
  border-radius: var(--radius);
  border: var(--border);
  background-color: var(--background-control);
  color: var(--color-meta);
}

.textarea .textarea__helper-text,
.textarea .textarea__counter {
  font-weight: var(--telekom-typography-font-weight-bold);
}

.textarea .textarea__counter {
  display: flex;
  transition: var(--transition-counter);
  margin-left: auto;
  padding-right: var(--telekom-spacing-unit-x3);
  justify-content: flex-end;
  font-size: var(--telekom-typography-font-size-small);
  line-height: var(--telekom-typography-line-spacing-standard);
  color: inherit;
}
.textarea .textarea__helper-text {
  transition: var(--transition-helper-text);
  padding-left: var(--spacing-x);
  font-size: var(--font-size-helper-text);
  line-height: var(--line-height-helper-text);
  color: var(--color-helper-text);
}
.textarea .textarea__meta {
  display: flex;
  justify-content: space-between;
  margin-top: var(--spacing-y-meta);
  color: var(--color-meta);
}

.textarea:not(.textarea--disabled):not(.textarea--readonly)
  .textarea__control:hover {
  border-color: var(--border-color-hover);
}

.textarea:not(.textarea--disabled):not(.textarea--readonly)
  .textarea__control:focus {
  border-color: var(--border-color-focus);
  box-shadow: var(--box-shadow-focus);
}

.textarea:not(.textarea--disabled) .textarea__control:focus::placeholder {
  color: var(--color-placeholder);
}

.textarea .textarea__control::placeholder,
.textarea ::placeholder {
  color: transparent;
  transition: var(--transition-placeholder);
}

.textarea__label {
  top: 0;
  left: 0;
  color: var(--color-label);
  display: flex;
  z-index: var(--z-index-label);
  position: absolute;
  transition: var(--transition-label);
  pointer-events: none;
  font-size: var(--font-size-label);
  font-weight: var(--font-weight-label);
  transform: translate(
    var(--spacing-x),
    calc((var(--telekom-spacing-unit-x12) - var(--font-size-label)) / 2)
  );
}
.textarea--has-focus .textarea__label,
.animated .textarea__label {
  transform: translate(var(--spacing-x), var(--telekom-spacing-unit-x2));
  font-size: var(--font-size-label-focus);
  font-weight: var(--font-weight-label-focus);
}
.textarea--status-error .textarea__control {
  border: var(--border-error);
}
.textarea--status-error .textarea__helper-text {
  color: var(--color-helper-text-error);
}
.textarea--status-error .textarea__counter {
  color: var(--color-counter-error);
}
.textarea--transparent .textarea__control {
  background-color: transparent;
}
.textarea__label-safety-background {
  top: var(--telekom-spacing-unit-x05);
  left: var(--telekom-spacing-unit-x05);
  right: var(--telekom-spacing-unit-x05);
  position: absolute;
  pointer-events: none;
  border-radius: var(--telekom-radius-standard);
  height: var(--telekom-spacing-unit-x6);
  background-color: var(--telekom-color-ui-base, #ffffff);
}
.textarea--transparent .textarea__label-safety-background,
.textarea--readonly .textarea__label-safety-background,
.textarea--disabled .textarea__label-safety-background {
  background-color: transparent;
}

.textarea--readonly textarea,
.textarea--readonly .textarea__control {
  color: var(--color-label-readonly);
  border-color: var(--border-color-readonly);
  background: var(--background-readonly);
}

.textarea--readonly .textarea__control:focus {
  box-shadow: var(--box-shadow-focus);
}

.textarea--disabled label,
.textarea--disabled .textarea__label,
.textarea--disabled textarea,
.textarea--disabled .textarea__control,
.textarea--disabled .textarea__meta,
.textarea--disabled .textarea__helper-text {
  cursor: not-allowed;
  border-color: var(--border-color-disabled);
  color: var(--color-disabled);
  background: transparent;
}
