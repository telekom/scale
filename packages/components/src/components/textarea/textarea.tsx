/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import {
  Component,
  Prop,
  Event,
  h,
  EventEmitter,
  Host,
  State,
} from '@stencil/core';
import classNames from 'classnames';

interface InputChangeEventDetail {
  value: string | number | boolean | undefined | null;
}

let i = 0;

@Component({
  tag: 'scale-textarea',
  styleUrl: './textarea.css',
  shadow: false,
})
export class Textarea {
  /** (optional) Input name */
  @Prop() name?: string = '';
  /** (optional) Input label */
  @Prop() label: string = '';
  /** (optional) textarea row */
  @Prop() rows?: number;
  /** (optional) textarea column */
  @Prop() cols?: number;
  /** (optional) Input helper text */
  @Prop() helperText?: string = '';
  /** (optional) Input status */
  @Prop() status?: string = '';
  /** (optional) Input max length */
  @Prop() maxLength?: number;
  /** (optional) Input min length */
  @Prop() minLength?: number;
  /** (optional) Input placeHolder */
  @Prop() placeholder?: string = '';
  /** (optional) Input disabled */
  @Prop() disabled?: boolean;
  /** (optional) Input readonly */
  @Prop() readonly?: boolean;
  /** (optional) Input required */
  @Prop() required?: boolean;
  /** (optional) Input counter */
  @Prop() counter?: boolean;
  /** (optional) textarea resize */
  @Prop() resize?: 'unset' | 'none' | 'vertical' | 'horizontal';
  /** (optional) Input value */
  @Prop({ mutable: true }) value?: string | number | null = '';
  /** (optional) Input checkbox id */
  @Prop() inputId?: string;
  /** (optional) input background transparent */
  @Prop() transparent?: boolean;
  /** (optional) Injected CSS styles */
  @Prop() styles?: string;

  /** Emitted when a keyboard input occurred. */
  @Event() scaleInput!: EventEmitter<KeyboardEvent>;
  /** Emitted when the value has changed. */
  @Event() scaleChange!: EventEmitter<InputChangeEventDetail>;
  /** Emitted when the input has focus. */
  @Event() scaleFocus!: EventEmitter<void>;
  /** Emitted when the input loses focus. */
  @Event() scaleBlur!: EventEmitter<void>;
  /** Emitted on keydown. */
  @Event() scaleKeyDown!: EventEmitter<KeyboardEvent>;

  /** Whether the input element has focus */
  @State() hasFocus: boolean = false;

  componentWillLoad() {
    if (this.inputId == null) {
      this.inputId = 'input-textarea' + i++;
    }
  }

  // We're not watching `value` like we used to
  // because we get unwanted `scaleChange` events
  // because how we keep this.value up-to-date for type="select"
  // `this.value = selectedValue`
  emitChange() {
    this.scaleChange.emit({
      value: this.value == null ? this.value : this.value.toString(),
    });
  }

  handleInput = (event: Event) => {
    const target = event.target as HTMLInputElement | null;
    if (target) {
      this.value = target.value || '';
      this.emitChange();
    }
    this.scaleInput.emit(event as KeyboardEvent);
  };

  handleChange = (event: Event) => {
    const target = event.target as HTMLInputElement | null;
    if (target) {
      this.value = target.value || '';
      this.emitChange();
    }
  };

  handleFocus = () => {
    this.scaleFocus.emit();
    this.hasFocus = true;
  };

  handleBlur = () => {
    this.scaleBlur.emit();
    this.hasFocus = false;
  };

  handleKeyDown = (event: KeyboardEvent) => {
    this.scaleKeyDown.emit(event);
  };

  render() {
    const ariaInvalidAttr =
      this.status === 'error' ? { 'aria-invalid': true } : {};
    const helperTextId = `helper-message-${i}`;
    const ariaDescribedByAttr = { 'aria-describedBy': helperTextId };
    const readonlyAttr = this.readonly ? { readonly: 'readonly' } : {};

    return (
      <Host>
        <div class={this.getCssClassMap()}>
          {/* Accessibility: label should be always *before* the actual input */}
          <label class="textarea__label" htmlFor={this.inputId}>
            {this.label}
          </label>
          <textarea
            class="textarea__control"
            style={!!this.resize && { resize: this.resize }}
            value={this.value}
            {...(!!this.name ? { name: this.name } : {})}
            required={this.required}
            minLength={this.minLength}
            maxLength={this.maxLength}
            id={this.inputId}
            onInput={this.handleInput}
            onChange={this.handleChange}
            onFocus={this.handleFocus}
            onBlur={this.handleBlur}
            onKeyDown={this.handleKeyDown}
            {...(!!this.placeholder ? { placeholder: this.placeholder } : {})}
            disabled={this.disabled}
            {...readonlyAttr}
            {...(!!this.rows ? { rows: this.rows } : {})}
            {...(!!this.cols ? { cols: this.cols } : {})}
            {...ariaInvalidAttr}
            {...(this.helperText ? ariaDescribedByAttr : {})}
          />

          {/* Accessibility: solid background for the textarea label to avoid making the label unreadable when there's text underneath */}
          <span class="textarea__label-safety-background" aria-hidden="true" />
          {(!!this.helperText || !!this.counter) && (
            <div
              class="textarea__meta"
              id={helperTextId}
              aria-live="polite"
              aria-relevant="additions removals"
            >
              {!!this.helperText && (
                <div class="textarea__helper-text">{this.helperText}</div>
              )}
              {this.counter && (
                <div class="textarea__counter">
                  {!!this.value ? String(this.value).length : 0} /{' '}
                  {this.maxLength}
                </div>
              )}
            </div>
          )}
        </div>
      </Host>
    );
  }

  getCssClassMap() {
    return classNames(
      'textarea',
      this.hasFocus && 'textarea--has-focus',
      this.resize && `textarea--resize-${this.resize}`,
      this.disabled && `textarea--disabled`,
      this.transparent && 'textarea--transparent',
      this.status && `textarea--status-${this.status}`,
      this.value != null && this.value !== '' && 'animated'
    );
  }
}
