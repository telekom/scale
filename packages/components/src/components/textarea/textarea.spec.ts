/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { newSpecPage } from '@stencil/core/testing';
import { Textarea } from './textarea';
describe('Textarea', () => {
  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [Textarea],
      html: `<scale-textarea></scale-textarea>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should match snapshot -> helpertext', async () => {
    const page = await newSpecPage({
      components: [Textarea],
      html: `
        <scale-textarea
          status="error"
          helper-text="helpertext"
          >
        </scale-textarea>`,
    });
    expect(page.root).toMatchSnapshot();
  });
  it('should match snapshot -> counter', async () => {
    const page = await newSpecPage({
      components: [Textarea],
      html: `
        <scale-textarea
          counter="true"
          >
        </scale-textarea>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  describe('classes', () => {
    it('should handle getCssClassMap()', () => {
      const element = new Textarea();
      expect(element.getCssClassMap()).toContain('textarea');

      element.hasFocus = true;
      expect(element.getCssClassMap()).toContain('textarea--has-focus');

      element.resize = 'vertical';
      expect(element.getCssClassMap()).toContain('textarea--resize-vertical');

      element.disabled = true;
      expect(element.getCssClassMap()).toContain('textarea--disabled');

      element.transparent = true;
      expect(element.getCssClassMap()).toContain('textarea--transparent');

      element.status = 'error';
      expect(element.getCssClassMap()).toContain('textarea--status-error');

      element.value = 'value';
      expect(element.getCssClassMap()).toContain('animated');
    });
  });
  describe('props', () => {
    it('check default props', async () => {
      const page = await newSpecPage({
        components: [Textarea],
        html: `<scale-textarea></scale-textarea>`,
      });
      expect(page.rootInstance.name).toBe('');
      expect(page.rootInstance.label).toBe('');
      expect(page.rootInstance.helperText).toBe('');
      expect(page.rootInstance.status).toBe('');
      expect(page.rootInstance.placeholder).toBe('');
      expect(page.rootInstance.value).toBe('');
    });

    it('check props being set', async () => {
      const page = await newSpecPage({
        components: [Textarea],
        html: `<scale-textarea></scale-textarea>`,
      });
      page.root.name = 'name';
      page.root.label = 'label';
      page.root.rows = 5;
      page.root.cols = 3;
      page.root.maxLength = 100;
      page.root.minLength = 1;
      page.root.placeholder = 'placeholder';
      page.root.required = true;
      page.root.counter = true;
      page.root.resize = 'vertical';
      page.root.disabled = 'true';
      page.root.transparent = 'true';
      page.root.value = 'value';
      page.root.inputId = 'inputId';
      page.root.styles = 'background : red';
      await page.waitForChanges();
      expect(page.rootInstance.name).toBe('name');
      expect(page.rootInstance.label).toBe('label');
      expect(page.rootInstance.rows).toBe(5);
      expect(page.rootInstance.cols).toBe(3);
      expect(page.rootInstance.maxLength).toBe(100);
      expect(page.rootInstance.minLength).toBe(1);
      expect(page.rootInstance.placeholder).toBe('placeholder');
      expect(page.rootInstance.required).toBe(true);
      expect(page.rootInstance.counter).toBe(true);
      expect(page.rootInstance.resize).toBe('vertical');
      expect(page.rootInstance.disabled).toBe(true);
      expect(page.rootInstance.value).toBe('value');
      expect(page.rootInstance.transparent).toBe(true);
      expect(page.rootInstance.inputId).toBe('inputId');
      expect(page.rootInstance.styles).toBe('background : red');
    });
  });
  describe('functions', () => {
    it('handleFocus() sets focus to true', () => {
      const element = new Textarea();
      element.handleFocus();
      expect(element.hasFocus).toBe(true);
    });

    it('handleBlur() sets focus to false', () => {
      const element = new Textarea();
      element.hasFocus = true;
      element.handleBlur();
      expect(element.hasFocus).toBe(false);
    });
  });
  describe('events', () => {
    let page;
    beforeEach(async () => {
      page = await newSpecPage({
        components: [Textarea],
        html: `<scale-textarea></scale-textarea>`,
      });
    });

    it('onKeyDown', async () => {
      const mock = jest.fn();
      page.root.addEventListener('scaleKeyDown', mock);
      const target = page.root.querySelector('.textarea__control');
      await page.waitForChanges();
      target.dispatchEvent(new KeyboardEvent('keydown'));
      await page.waitForChanges();
      expect(mock).toHaveBeenCalled();
    });

    it('onBlur', async () => {
      const mock = jest.fn();
      page.root.addEventListener('scaleBlur', mock);
      const target = page.root.querySelector('.textarea__control');
      await page.waitForChanges();
      target.dispatchEvent(new Event('blur'));
      await page.waitForChanges();
      expect(mock).toHaveBeenCalled();
    });

    it('onFocus', async () => {
      const mock = jest.fn();
      page.root.addEventListener('scaleFocus', mock);
      const target = page.root.querySelector('.textarea__control');
      await page.waitForChanges();
      target.dispatchEvent(new Event('focus'));
      await page.waitForChanges();
      expect(mock).toHaveBeenCalled();
    });

    it('onInput', async () => {
      const mock = jest.fn();
      page.root.addEventListener('scaleInput', mock);
      const target = page.root.querySelector('.textarea__control');
      await page.waitForChanges();
      target.dispatchEvent(new Event('input'));
      await page.waitForChanges();
      expect(mock).toHaveBeenCalled();
    });

    it('onChange', async () => {
      const mock = jest.fn();
      page.root.emitChange = jest.fn();
      page.root.addEventListener('scaleChange', mock);
      const target = page.root.querySelector('.textarea__control');
      await page.waitForChanges();
      target.dispatchEvent(new Event('change'));
      await page.waitForChanges();
      expect(mock).toHaveBeenCalled();
      expect(page.root.emitChange).not.toHaveBeenCalled();
    });
  });
  describe('branches', () => {
    let page;
    beforeEach(async () => {
      page = await newSpecPage({
        components: [Textarea],
        html: `<scale-textarea></scale-textarea>`,
      });
    });
    it('...', async () => {
      const mock = jest.fn();
      page.root.addEventListener('scaleChange', mock);
      page.root.value = null;
      await page.waitForChanges();
      page.rootInstance.emitChange();
      await page.waitForChanges();
      expect(mock).toHaveBeenCalled();
    });
  });
});
