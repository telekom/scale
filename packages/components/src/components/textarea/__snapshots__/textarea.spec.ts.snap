// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea should match snapshot -> counter 1`] = `
<scale-textarea counter="true">
  <div class="textarea">
    <label class="textarea__label" htmlfor="input-textarea2"></label><textarea class="textarea__control" id="input-textarea2" value=""></textarea>
    <span aria-hidden="true" class="textarea__label-safety-background"></span>
    <div aria-live="polite" aria-relevant="additions removals" class="textarea__meta" id="helper-message-3">
      <div class="textarea__counter">
        0 /
      </div>
    </div>
  </div>
</scale-textarea>
`;

exports[`Textarea should match snapshot -> helpertext 1`] = `
<scale-textarea helper-text="helpertext" status="error">
  <div class="textarea textarea--status-error">
    <label class="textarea__label" htmlfor="input-textarea1"></label><textarea aria-describedby="helper-message-2" aria-invalid="" class="textarea__control" id="input-textarea1" value=""></textarea>
    <span aria-hidden="true" class="textarea__label-safety-background"></span>
    <div aria-live="polite" aria-relevant="additions removals" class="textarea__meta" id="helper-message-2">
      <div class="textarea__helper-text">
        helpertext
      </div>
    </div>
  </div>
</scale-textarea>
`;

exports[`Textarea should match snapshot 1`] = `
<scale-textarea>
  <div class="textarea">
    <label class="textarea__label" htmlfor="input-textarea0"></label><textarea class="textarea__control" id="input-textarea0" value=""></textarea>
    <span aria-hidden="true" class="textarea__label-safety-background"></span>
  </div>
</scale-textarea>
`;
